src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Float.cs

```diff
@@ -88,9 +89,71 @@ private void WriteNumberValueIndented(float value)
                 BytesPending += indent;
             }
 
-            bool result = Utf8Formatter.TryFormat(value, output.Slice(BytesPending), out int bytesWritten);
+            bool result = TryFormatSingle(value, output.Slice(BytesPending), out int bytesWritten);
             Debug.Assert(result);
             BytesPending += bytesWritten;
         }
+
+        private static bool TryFormatSingle(float value, Span<byte> destination, out int bytesWritten)
+        {
+            // Frameworks that are not .NET Core 3.0 or higher do not produce roundtrippable strings by
+            // default. Further, the Utf8Formatter on older frameworks does not support taking a precision
+            // specifier for 'G' nor does it represent other formats such as 'R'. As such, we duplicate
+            // the .NET Core 3.0 logic of forwarding to the UTF16 formatter and transcoding it back to UTF8.
+
+#if BUILDING_INBOX_LIBRARY
+            return Utf8Formatter.TryFormat(value, destination, out bytesWritten);
+#else
+            const string FormatString = "G9";
+
+            // We first try to format into a stack-allocated buffer, and if it succeeds, we can avoid
+            // all allocation.  If that fails, we fall back to allocating strings.  If it proves impactful,
+            // that allocation (as well as roundtripping from byte to char and back to byte) could be avoided by
+            // calling into a refactored Number.FormatSingle/Double directly.
+
+            const int StackBufferLength = 128; // large enough to handle the majority cases
+            Span<char> stackBuffer = stackalloc char[StackBufferLength];
+            ReadOnlySpan<char> utf16Text = stackalloc char[0];
+
+            // Try to format into the stack buffer.  If we're successful, we can avoid all allocations.
+            if (value.TryFormat(stackBuffer, out int formattedLength, FormatString, CultureInfo.InvariantCulture))
```

Comment:
Can you share the code coverage numbers for this code path? Are we excising all the code paths here?

Reply:
Can you share how to get code coverage numbers?

I would also strongly push back against modifying this at all. It was taken as is, modulo the conversion of `StandardFormat` to the utf-16 format string, from the Utf8Formatter.

Reply:
The "hacky" way to do it would be to flip (or remove) the .NET Core 3.0 code path/ifdef and then run the tests with `/p:Coverage=true`. This way, for all platforms the custom code runs.

I am not sure how to get coverage for non-netcoreapp build configuration.

cc @ViktorHofer

Reply:
Code coverage is only supported on netcoreapp. I think manually switching the ifdefs to collect coverage is fine